src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WinHttpWebSocket.cs

```diff
@@ -441,6 +448,12 @@ private Task<bool> InternalReceiveAsync(ArraySegment<byte> buffer)
                 Abort();
             }
 
+            // If the task has completed, there's no need to force continuations to run async
+            if (_operation.TcsReceive.Task.Status == TaskStatus.RanToCompletion)
+            {
+                return Task.FromResult(_operation.TcsReceive.Task.GetAwaiter().GetResult());
+            }
```

Comment:
@mconnew, I think we're still taking past each other. I'm not asking about your RunContinuationsAsynchronously change; that's fine. I'm asking about this if block I commented on. If by the time you get to this if block the operation hasn't completed, the if block is irrelevant. If it has completed, then with or without the if block you're returning a completed task. Since it's completed by the time this method returns, awaits at the call site will behave exactly the same whether you return the tcs task or CompletedTask, since they're both already successfully completed. Hence my question about why you have this if block and what use case it's impacting. I can think of only two reasons: you're using ContinueWith(..., ExrcuteSynchronously) at the call site (that would behave differently for two successfully completed tasks, one with RunContinuationsAsynchronously and one without), or you've noticed some minute perf difference from letting the tcs task get gc'd slightly earlier. Both of those seem unlikely.